osbuild/remoteloop.py

```diff
@@ -63,6 +63,33 @@ def _create_device(self, fd, dir_fd, offset=None, sizelimit=None):
                 lo.close()
                 continue
             break
+        return lo
+
+    def _create_device(self, fd, dir_fd, offset=None, sizelimit=None):
+        while True:
+            # Getting an unbound loopback device and attaching a backing
+            # file descriptor to it is racy, so we must use a retry loop
+            lo = loop.Loop(self.ctl.get_unbound())
+            try:
+                lo.configure(fd, offset=offset, sizelimit=sizelimit, autoclear=True)
+            except BlockingIOError:
+                lo.clear_fd()
+                lo.close()
+                continue
+            except OSError as e:
+                lo.close()
+                # `configure` returns EBUSY when the pages from the previously
+                # bound file have not been fully cleared yet.
+                if e.errno == errno.EBUSY:
+                    continue
+                if e.errno == errno.ENOTTY:
+                    lo = None
+                    break
+                raise e
+            break
+
+        if not lo:
+            lo = self._legacy_create_device(fd, offset, sizelimit)
```

Comment:
> In addition, you could then test in unit test each of this variants of _create_device() separately based on the kernel version. However note that the kernel version check may get tricky, since you can not be sure the LOOP_CONFIGURE won't be available on old RHEL kernel versions, since the functionality may be backported.

interesting, it might already be the case, because CI was passing while I had a bug in the fallback.
I think it's not really an issue, because unit tests are for us and the CI checks. Which can easily run on an updated machine. However we may never properly test the fallback.

So I'm wondering, can we simply get rid of the fallback ? If every distributions we support has the `ioctl` implemented, we should be covered. No ?

Reply:
> So I'm wondering, can we simply get rid of the fallback ? If every distributions we support has the `ioctl` implemented, we should be covered. No ?

RHEL-8 still uses kernel 4.18.0. Unless you can prove that the feature is available on all distributions that we need to support, I suggest to take the safe path and keep the fallback.

Reply:
> > So I'm wondering, can we simply get rid of the fallback ? If every distributions we support has the `ioctl` implemented, we should be covered. No ?
> 
> RHEL-8 still uses kernel 4.18.0. Unless you can prove that the feature is available on all distributions that we need to support, I suggest to take the safe path and keep the fallback.

I can try running CI without the fallback and see it that works ?

Reply:
Our CI is not comprehensive. It does not test on RHEL at all and also does not test on other architectures than x86_64.

Reply:
So potentially yes, but I incline to keeping the fallback in the code for now.

Reply:
ftr, `LOOP_CONFIGURE` was backported into RHEL 8 kernel in RHEL 8.4 as a part of https://bugzilla.redhat.com/show_bug.cgi?id=1881760 (block layer: update to upstream v5.8).

Since RHEL 8.4 is currently the oldest supported release that we support running `osbuild` on, it might be just fine implementing this without the fallback...

From a centos stream 8 container:

```
# rpm -q kernel
kernel-4.18.0-448.el8.x86_64
# rpm -q --changelog kernel | grep LOOP_CONFIGURE
- loop: Fix missing discard support when using LOOP_CONFIGURE (Ming Lei) [1997338]
- [block] loop: Set correct device size when using LOOP_CONFIGURE (Ming Lei) [1881760]
- [block] loop: unset GENHD_FL_NO_PART_SCAN on LOOP_CONFIGURE (Ming Lei) [1881760]
- [block] loop: Add LOOP_CONFIGURE ioctl (Ming Lei) [1881760]
- ```

Reply:
I'm fine with that @ondrejbudai and thanks for providing this info.

I expected @lavocatt to provide this kind of information to support his suggestion of using only `LOOP_CONFIGURE`.

Reply:
Thanks @ondrejbudai. To be honest, I would probably not have found this information. I'm not yet versed enough into bugzilla and bug navigation in general, (working on getting up to speed in any case). I also have a `You are not authorized to access bug #1881760.` error following on the link, even logged in as a Red Hat associate.